core/src/main/java/org/elasticsearch/action/admin/indices/analyze/TransportAnalyzeAction.java

```diff
@@ -470,27 +422,156 @@ private void analyze(TokenStream stream, Analyzer analyzer, String field, Set<St
         final Map<String, Object> extendedAttributes = new TreeMap<>();
 
-        stream.reflectWith(new AttributeReflector() {
-            @Override
-            public void reflect(Class<? extends Attribute> attClass, String key, Object value) {
-                if (CharTermAttribute.class.isAssignableFrom(attClass))
-                    return;
-                if (PositionIncrementAttribute.class.isAssignableFrom(attClass))
-                    return;
-                if (OffsetAttribute.class.isAssignableFrom(attClass))
-                    return;
-                if (TypeAttribute.class.isAssignableFrom(attClass))
-                    return;
-                if (includeAttributes == null || includeAttributes.isEmpty() || includeAttributes.contains(key.toLowerCase(Locale.ROOT))) {
-                    if (value instanceof BytesRef) {
-                        final BytesRef p = (BytesRef) value;
-                        value = p.toString();
-                    }
-                    extendedAttributes.put(key, value);
+        stream.reflectWith((attClass, key, value) -> {
+            if (CharTermAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (PositionIncrementAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (OffsetAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (TypeAttribute.class.isAssignableFrom(attClass))
+                return;
+            if (includeAttributes == null || includeAttributes.isEmpty() || includeAttributes.contains(key.toLowerCase(Locale.ROOT))) {
+                if (value instanceof BytesRef) {
+                    final BytesRef p = (BytesRef) value;
+                    value = p.toString();
                 }
+                extendedAttributes.put(key, value);
             }
         });
 
         return extendedAttributes;
     }
+
+    private static CharFilterFactory[] getCharFilterFactories(AnalyzeRequest request, AnalysisService analysisService, AnalysisRegistry analysisRegistry, Environment environment, CharFilterFactory[] charFilterFactories) throws IOException{
+        if (request.charFilters() != null && request.charFilters().length > 0) {
+            charFilterFactories = new CharFilterFactory[request.charFilters().length];
+            for (int i = 0; i < request.charFilters().length; i++) {
+                final String charFilterName = request.charFilters()[i];
+                // parse custom settings if a value start "{"
+                if (charFilterName.startsWith("{")) {
+                    Settings settings = getCustomSettings(charFilterName);
+                    String charFilterTypeName = settings.get("type");
+                    if (charFilterTypeName == null) {
+                        throw new IllegalArgumentException("char filter type does not defined. [" + charFilterName + "]");
```

Comment:
It would be good to have some context as to where the filter appeared. I would at least word it a little differently:
```"Missing [type] setting for anonymous char filter: " + charFilterName```